c/xudt_rce.c

```diff
@@ -0,0 +1,526 @@
+
+// uncomment to enable printf in CKB-VM
+//#define CKB_C_STDLIB_PRINTF
+//#include <stdio.h>
+
+// it's used by blockchain-api2.h, the behavior when panic
+#ifndef MOL2_EXIT
+#define MOL2_EXIT ckb_exit
+#endif
+int ckb_exit(signed char);
+
+#include <stdbool.h>
+#include <string.h>
+
+#include "blake2b.h"
+#include "blockchain-api2.h"
+#include "ckb_consts.h"
+
+#if defined(CKB_USE_SIM)
+#include <stdio.h>
+
+#include "ckb_syscall_xudt_sim.h"
+#define xudt_printf printf
+#else
+// it will be re-defined in ckb_dlfcn.h
+#undef MAX
+#undef MIN
+#include "ckb_dlfcn.h"
+#include "ckb_syscalls.h"
+#define xudt_printf(x, ...) (void)0
+#endif
+
+#define BLAKE160_SIZE 20
+#define SCRIPT_SIZE 32768
+#define RAW_EXTENSION_SIZE 65536
+#define EXPORTED_FUNC_NAME "validate"
+#define MAX_CODE_SIZE (1024 * 1024)
+#define FLAGS_SIZE 4
+#define MAX_LOCK_SCRIPT_HASH_COUNT 2048
+
+#include "rce.h"
+
+// global variables, type definitions, etc
+
+// We will leverage gcc's 128-bit integer extension here for number crunching.
+typedef unsigned __int128 uint128_t;
+
+uint8_t g_script[SCRIPT_SIZE] = {0};
+uint8_t g_raw_extension_data[RAW_EXTENSION_SIZE] = {0};
+WitnessArgsType g_witness_args;
+
+/*
+is_owner_mode indicates if current xUDT is unlocked via owner mode(as
+described by sUDT), extension_index refers to the index of current extension in
+the ScriptVec structure. args and args_length are set to the script args
+included in Script structure of current extension script.
+
+If this function returns 0, the validation for current extension script is
+consider successful.
+ */
+typedef int (*ValidateFuncType)(int is_owner_mode, size_t extension_index,
+                                const uint8_t* args, size_t args_len);
+
+typedef enum XUDTFlags {
+  XUDTFlagsPlain = 0,
+  XUDTFlagsInArgs = 1,
+  XUDTFlagsInWitness = 2,
+} XUDTFlags;
+
+typedef enum XUDTValidateFuncCategory {
+  CateNormal = 0,  // normal extension script
+  CateRce = 1,     // Regulation Compliance Extension
+} XUDTValidateFuncCategory;
+
+uint8_t RCE_HASH[32] = {1};
```

Comment:
The RCE pre-defined code hash in [RFC](https://talk.nervos.org/t/rfc-extensible-udt/5337) is ` 0x0000 ... 0001` but here the value is `0x0100 ... 0000`.

Need confirmation from @xxuejie .

Reply:
The original definition is indeed `0x0000 ... 0001`, but we don't have strong feelings either way.

Reply:
in RFC, it's said  "for example", it's OK to defined in code:
"Some extension logics might have predefined hash, for example, we can use 0x0000 ... 0001 to represent regulation extension. The actual code for such scripts can be embedded in xUDT script itself."